Hadoop LZ4 Support for LZ4 Codec

marioloko opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #2988 .

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

It adds support for `LZ4_HADOOP` compression for `CompressionCodec::LZ4`, with fallback to `LZ4_FRAME` and `LZ4_RAW` algorithm for backward compatibility with older `arrow-rs` versions and older `parquet-cpp` versions, respectively.

It also adds tests for `LZ4_HADOOP` files, and tests for the fallback to `LZ4_RAW` (a test including 2 files with same content, same `CompressionCodec::LZ4`, but different compression algorithm `LZ4_HADOOP` and `LZ4_RAW`).

From now on, `CompressionCodec::LZ4` will compress files using `LZ4_HADOOP` algorithm, making LZ4 files generated by this library compatible with other languages arrow implementations.

The user interface is changed to allow the user to configure whether to fallback or not, by allowing the user to disable backward compatibility with `no_backward_compatible_lz4` (WIP in this PR).

# Are there any user-facing changes?

The user interface is changed to allow the user to configure whether to fallback or not, by allowing the user to disable backward compatibility with `no_backward_compatible_lz4` (WIP in this PR).

New structs have been created:
`SerializedRowGroupReaderOptions`and `SerializedPageReaderOptions` to allow the users to set the `CodecOptions` for `SerializedRowGroupReader` and `SerializedPageReader`. The builders for these structs are also added to public interface. 
`SerializedRowGroupReader::new_with_options` and `SerializedPageReader::new_with_options` have been created to allow to configure these components without breaking public API.

NOTE: The API changes are WIP, however, it is not possible to follow the same approach that with `WriterProperties` and wrap `ReadOptions` in an `Arc` as its field `predicates: Vec<ReadGroupPredicate>` is not thread safe. Other option would be to wrap it in just an `Rc` so `type ReadOptionsPtr = Rc<ReadOptions>`.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

marioloko commented:
Another thing I notice with using `ReadOptionsPtr` is that even if we share the ownership of the pointer we need to take ownership of some of its values here:
```
    pub fn new_with_options(chunk_reader: R, options: ReadOptions) -> Result<Self> {
        let metadata = footer::parse_metadata(&chunk_reader)?;
        let mut predicates = options.predicates;
```

Of course these options are not used anymore at this point, so instead of cloning the vector of predicates we can just take it.
```
    pub fn new_with_options(chunk_reader: R, options: ReadOptions) -> Result<Self> {
        let metadata = footer::parse_metadata(&chunk_reader)?;
        let mut predicates = std::men::take(&mut options.predicates);
```

marioloko commented:
If we want to hide the `CodecOptions` probably we have to create a `ReadProperties` struct. Which in this case is only a wrap around `CodecOptions` but different config could be included, the restrictions for fields in this struct is to be `Send` and `Sync`, and to be read only. In contrast with `ReadOptions` which consumes the fields and is not `Send` nor `Sync`.

This problem also implies implementing setters multiple times, but we can use the same `ReadPropertiesPtr` for `SerializedPageReader` and `SerializedRowGroupReader`.

@tustvold what do you think?

tustvold commented:
Sorry for the delay, I'm currently on holiday. Will try to find time to review today

marioloko commented:
Oh, do not worry, enjoy your holidays! We review this when you are back 😄

marioloko commented:
With this last change we remove `CodecOptions` from public API by wrapping it into a more general `ReaderProperties`.

# Commit description:

This commits hides `CodecOptions` from the public API. The changes are the following:
- Added a new structs to public API `ReaderProperties`, `ReaderPropertiesBuilder` and `ReaderPropertiesPtr` to store inmutable reader config, as it is the case of `CodecOptions`.
- Removed `SerializedRowGroupReaderOptions`, `SerializedRowGroupReaderOptionsBuilder`, `SerializedPageReaderOptionsBuilder` and `SerializedPageReaderOptions`. They are not required anymore as `SerializedRowGroupReader` and `SerializedRowGroupReaderOptions` use `ReaderPropertiesPtr` for configuration.
- `SerializedRowGroupReader::new_with_options` renamed to `SerializedRowGroupReader::new_with_properties`.
- `SerializedPageReader::new_with_options` renamed to `SerializedPageReader::new_with_properties`.
- Test added for `ReaderPropertiesBuilder`.

marioloko commented:
I think that now it is everything is ready to merge. Take a look when you have time and we discuss if any change has to be done.

About the failing test above I think it is not caused by my commit, as I have not modify IPC. In fact, in my local Mac the test is passing.

tustvold commented on parquet/src/arrow/arrow_reader/mod.rs:
```diff
@@ -2422,6 +2422,76 @@ mod tests {
         assert_eq!(a.values(), &[42.000000, 7.700000, 42.125000, 7.700000]);
     }
 
+    // This test is to ensure backward compatibility, it test 2 files containing the LZ4 CompressionCodec
```
Love the tests

tustvold commented on parquet/src/compression.rs:
```diff
@@ -414,15 +472,202 @@ mod lz4_raw_codec {
+    }
+
+    /// Try to decompress the buffer as if it was compressed with the Hadoop Lz4Codec.
+    /// Adapted from pola-rs [compression.rs:try_decompress_hadoop](https://pola-rs.github.io/polars/src/parquet2/compression.rs.html#225)
```
:+1: nice to see attribution

tustvold reviewed (approved):
I will give this another once over prior to merge, but looking good. Thank you for your work on this 👍

tustvold commented:
Thanks again :smile:

tustvold commented:
Marking as api-change even though backwards compatible as does change the output, and so should be highlighted in changelog
